File: 09.19/프로그래머스/조수연/k진수에서 소수 개수 구하기.kt
package com.ssafy.algorithm

fun main() {
    val sl = Solution()
    val n = 110011
    val k = 10
    println(sl.solution(n, k))
}

class Solution {
    var answer: Int = 0
    fun solution(n: Int, k: Int): Int {
        var num = n
        val sb = StringBuilder()
        while (true) { //진수로 변환하기
            if (num / k == 0) {
                sb.insert(0, num % k)
                break
            }
            sb.insert(0, num % k)
            num /= k
        }
        val code = sb.toString()

        val nums = code.split("0")

        check(nums) //소수 찾기기

        return answer
    }

    fun check(nums: List<String>) {
        for (num in nums) {
            if (num != "") { //split 잘랐을 때 ""이 리스트에 담길수도 있음 제외하고 탐색
                val prime = num.toLong() 
                if (prime != 1L) { //1은 pass
                    if (isPrime(prime)) answer++
                }
            }
        }
    }

    fun isPrime(prime: Long): Boolean { // 소수 판별

        for (i in 2..Math.sqrt(prime.toDouble()).toInt()) { // 숫자의 제곱근까지 loop 해서 판별 
            if ((prime % i).toInt() == 0) return false
        }

        return true
    }
}

File: 2022/09.19/프로그래머스/이아름/k진수에서 소수개수 구하기.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

bool sosu(long long n) {
	if (n <= 1) return false;
	if (n == 2) return true;
	for (long long i = 3; i <= sqrt(n); i++) {
		if (n%i == 0) return false;
	}
	return true;
}

int solution(int n, int k) {
	int answer = 0;
	string str = "";
	while (n) {
		str = to_string(n % k) + str;
		n /= k;
	}
	string cut = "";
	for (int i = 0; i < str.size(); i++) {
		if (str[i] == '0') {
			if (!cut.empty() && sosu(stoi(cut))) {
				answer++;
			}
			cut = "";
		}
		else {
			cut += str[i];
		}
	}
	if (!cut.empty() && sosu(stol(cut))) {
		answer++;
	}
	return answer;
}
